job-runner/chain/traverser.go

```diff
@@ -377,3 +396,99 @@ func (t *traverser) sendJL(pJob proto.Job, err error) {
+
+	// sequenceJobsToRetry is a list containing the failed job and all previously
+	// completed jobs in the sequence. For example, if job C of A -> B -> C -> D
+	// fails, then A and B are the previously completed jobs and C is the failed
+	// job. So, jobs A, B, and C will be added to sequenceJobsToRetry. D will not be
+	// added because it was never ran.
+	var sequenceJobsToRetry []proto.Job
+	sequenceJobsToRetry = append(sequenceJobsToRetry, failedJob)
+	sequenceJobsCompleted := t.sequenceJobsCompleted(sequenceStartJob)
+	sequenceJobsToRetry = append(sequenceJobsCompleted, sequenceJobsCompleted...)
+
+	// Mark all jobs to retry to PENDING state
+	for _, job := range sequenceJobsToRetry {
+		t.chain.SetJobState(job.Id, proto.STATE_PENDING)
+	}
+
+	// Decrement number of jobs we will attempt to retry
+	t.chain.n -= uint(len(sequenceJobsToRetry))
+
+	// Increment retry count for this sequence
+	t.chain.IncrementSequenceRetryCount(failedJob)
+
+	// Enqueue first job in sequence
+	t.runJobChan <- sequenceStartJob
+
+	return nil
+}
+
+// sequenceJobsCompleted does a BFS to find all jobs in the sequence that have
+// completed. You can read how BFS works here:
+// https://en.wikipedia.org/wiki/Breadth-first_search.
+func (t *traverser) sequenceJobsCompleted(sequenceStartJob proto.Job) []proto.Job {
+	// toVisit is a map of job id->job to visit
+	toVisit := map[string]proto.Job{}
+	// visited is a map of job id->job visited
+	visited := map[string]proto.Job{}
+
+	// Process sequenceStartJob
+	for _, pJob := range t.chain.NextJobs(sequenceStartJob.Id) {
+		toVisit[pJob.Id] = pJob
+	}
+	visited[sequenceStartJob.Id] = sequenceStartJob
+
+PROCESS_TO_VISIT_LIST:
+	for len(toVisit) > 0 {
+
+	PROCESS_CURRENT_JOB:
+		for currentJobId, currentJob := range toVisit {
+
+		PROCESS_NEXT_JOBS:
+			for _, nextJob := range t.chain.NextJobs(currentJobId) {
+				// Don't add failed or pending jobs to toVisit list
+				// For example, if job C of A -> B -> C -> D fails, then do not add C
+				// or D to toVisit list. Because we have single sequence retries,
+				// stopping at the failed job ensures we do not add jobs not in the
+				// sequence to the toVisit list.
+				if nextJob.State != proto.STATE_COMPLETE {
+					continue PROCESS_NEXT_JOBS
+				}
+
+				// Make sure we don't visit a job multiple times. We can see a job
+				// mulitple times if it is a "fan in" node.
+				if _, seen := visited[nextJob.Id]; !seen {
+					toVisit[nextJob.Id] = nextJob
+				}
+			}
+
+			// Since we have processed all of the next jobs for this current job, we
+			// are done visiting the current job and can delete it from the toVisit
+			// list and add it to the visited list.
+			delete(toVisit, currentJobId)
+			visited[currentJobId] = currentJob
+
+			continue PROCESS_CURRENT_JOB
+		}
+
+		continue PROCESS_TO_VISIT_LIST
+	}
+
+	var completedJobs []proto.Job
+	for _, j := range visited {
+		completedJobs = append(completedJobs, j)
```

Comment:
Optimization: whenever we know the size of an array, we should allocate only that size:
```go
completedJobs := make([]proto.Job, 0, len(visited))
```
This prevents Go from doing dynamic allocations in bigger chunks that often go unused. Obviously not a huge pefr win, but it's easy enough to implement, so might as well.

Reply:
cool. good idea. updated it to be `make([]proto.Job, len(visited))` since we'll def fill the capacity.